src/playground-project.ts

```diff
@@ -546,6 +563,88 @@ export class PlaygroundProject extends LitElement {
     this.dispatchEvent(new CustomEvent('compileDone'));
   }
 
+  emptyCompletions() {
+    this._completions = [];
+  }
+
+  /**
+   * Query the language service for completion options on
+   * token under cursor in code-editor
+   * */
+  async updateCompletions(
+    filename: string,
+    fileContent: string,
+    tokenUnderCursor: EditorToken,
+    cursorIndex: number,
+    codeEditorChangeData: CodeEditorChangeData
+  ) {
+    const tokenUnderCursorAsString = tokenUnderCursor.string.trim();
+    // If the user is starting a new word, we need to fetch relevant completion items
+    // from the TypeScript Language Service. If we are however building on top of
+    // a already fetched completions list, by narrowing keyword matches, we can
+    // just work with what we have fetched earlier.
+    if (!codeEditorChangeData.isRefinement) {
+      const workerApi = await this._deferredTypeScriptWorkerApi.promise;
+      const completionInfo = await workerApi.getCompletions(
+        filename,
+        fileContent,
+        tokenUnderCursorAsString,
+        cursorIndex,
+        {importMap: this._importMap}
+      );
+      if (completionInfo) {
+        const getCompletionDetailsFunction =
+          this.getCompletionDetails.bind(this);
+        // We pre-generate the getter for each completion item's details, so that
+        // if neeeded, they can fetch their details themselves.
+        this._completionInfo = populateCompletionInfoWithDetailGetters(
+          completionInfo,
+          filename,
+          cursorIndex,
+          getCompletionDetailsFunction
+        );
+      }
+    }
+
+    const searchWordIsPeriod = tokenUnderCursor.string === '.';
+    // In the case that the search word is a period, we don't really
+    // have any material to fuzzy find with, so we don't have need
+    // for running the search results through a fuzzy search.
+    // For this case, we just return the entries as completion items as is.
+    if (searchWordIsPeriod) {
+      this._completions = completionEntriesAsEditorCompletions(
+        this._completionInfo?.entries,
+        '.'
+      );
+    } else {
+      this._completions = sortCompletionItems(
+        this._completionInfo?.entries,
+        tokenUnderCursorAsString
+      );
+    }
+    // We want to pre-fetch the first completion item, if it's present
+    // so that when the data gets to the code-editor, the detail hopefully
+    // is already loaded.
+    this._completions[0]?.details;
+    this.dispatchEvent(new CustomEvent('completionsChanged'));
+  }
+
+  async getCompletionDetails(
```

Comment:
Can `getCompletionDetails` be private? Looks like it's only referenced from within this class.